File: internal/dateutil/dateutil_test.go
package dateutil_test

import (
	"testing"
	"time"

	"actual_helper/internal/dateutil"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestDateutil(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Dateutil Suite")
}

var _ = Describe("FormatDate", func() {
	It("formats same month correctly", func() {
		stmtDate := time.Date(2026, time.July, 14, 0, 0, 0, 0, time.UTC)
		result, err := dateutil.FormatDate("15 JUN", stmtDate)
		Expect(err).NotTo(HaveOccurred())
		Expect(result).To(Equal("2026-06-15"))
	})

	It("formats previous year for December on January statement", func() {
		stmtDate := time.Date(2027, time.January, 14, 0, 0, 0, 0, time.UTC)
		result, err := dateutil.FormatDate("25 DEC", stmtDate)
		Expect(err).NotTo(HaveOccurred())
		Expect(result).To(Equal("2026-12-25"))
	})

	It("formats same month for January on January statement", func() {
		stmtDate := time.Date(2027, time.January, 14, 0, 0, 0, 0, time.UTC)
		result, err := dateutil.FormatDate("05 JAN", stmtDate)
		Expect(err).NotTo(HaveOccurred())
		Expect(result).To(Equal("2027-01-05"))
	})

	It("returns error for invalid format", func() {
		stmtDate := time.Date(2026, time.July, 14, 0, 0, 0, 0, time.UTC)
		_, err := dateutil.FormatDate("invalid", stmtDate)
		Expect(err).To(HaveOccurred())
	})

	It("returns error for invalid day", func() {
		stmtDate := time.Date(2026, time.July, 14, 0, 0, 0, 0, time.UTC)
		_, err := dateutil.FormatDate("XX JUL", stmtDate)
		Expect(err).To(HaveOccurred())
	})

	It("returns error for invalid month", func() {
		stmtDate := time.Date(2026, time.July, 14, 0, 0, 0, 0, time.UTC)
		_, err := dateutil.FormatDate("15 XXX", stmtDate)
		Expect(err).To(HaveOccurred())
	})
})

File: frontend/src/lib/__tests__/dateutil.test.ts
import { describe, it, expect } from 'vitest'
import { formatDate, parseTNGDate, parseDayMonthYear, parseStatementDate } from '../dateutil'

describe('dateutil', () => {
  describe('formatDate', () => {
    it('converts DD MMM to YYYY-MM-DD', () => {
      const stmtDate = new Date(2026, 0, 15) // Jan 15, 2026
      expect(formatDate('1 Jan', stmtDate)).toBe('2026-01-01')
      expect(formatDate('15 Dec', stmtDate)).toBe('2025-12-15')
    })
  })

  describe('parseTNGDate', () => {
    it('parses D/M/YYYY', () => {
      const d = parseTNGDate('1/5/2026')
      expect(d).not.toBeNull()
      expect(d!.getFullYear()).toBe(2026)
      expect(d!.getMonth()).toBe(4) // May = 4 (D/M format: 1 May 2026)
      expect(d!.getDate()).toBe(1)
    })

    it('parses DD/MM/YYYY', () => {
      const d = parseTNGDate('01/12/2026')
      expect(d).not.toBeNull()
      expect(d!.getMonth()).toBe(11) // Dec = 11
      expect(d!.getDate()).toBe(1)
    })
  })

  describe('parseDayMonthYear', () => {
    it('parses D January 2006', () => {
      const d = parseDayMonthYear('1 January 2026')
      expect(d).not.toBeNull()
      expect(d!.getFullYear()).toBe(2026)
      expect(d!.getMonth()).toBe(0)
      expect(d!.getDate()).toBe(1)
    })

    it('parses D Jan 2006', () => {
      const d = parseDayMonthYear('1 Jan 2026')
      expect(d).not.toBeNull()
      expect(d!.getFullYear()).toBe(2026)
    })
  })

  describe('parseStatementDate', () => {
    it('parses 02 Jan 2006', () => {
      const d = parseStatementDate('02 Jan 2026')
      expect(d).not.toBeNull()
      expect(d!.getFullYear()).toBe(2026)
      expect(d!.getMonth()).toBe(0)
      expect(d!.getDate()).toBe(2)
    })

    it('parses 02 Jan 06', () => {
      const d = parseStatementDate('02 Jan 06')
      expect(d).not.toBeNull()
      expect(d!.getFullYear()).toBe(2006)
    })
  })
})
